pages/DSNP/UserData.md

```diff
@@ -0,0 +1,92 @@
+DSNP User Data is designed for data with the following characteristics:
+* The data is intended to be accessible across the network with the strongest possible durability guarantees
+* The data may change frequently.
+* The data does not need to be widely announced and instead can be read just-in-time for a particular user.
+
+Data may be public or stored in encrypted form.
+The [Avro](https://avro.apache.org) schema specification is used to define the binary representation of relevant data types.
+
+## User Data Types
+
+DSNP implementations MUST support the following user data types:
+
+| System Name | Version | Encryption Algorithm | Compression Codec | Avro Object Type |
+| --- | --- | --- | --- | --- |
+| `publicFollows` | 1.2 | NONE | [`brotli`](https://brotli.org) | [GraphEdge](Graph.md#edges) |
+| `privateFollows` | 1.2 | `curve25519xsalsa20poly1305` |  [`brotli`](https://brotli.org) | [GraphEdge](Graph.md#edges) |
+| `privateConnections` | 1.2 | `curve25519xsalsa20poly1305` | [`brotli`](https://brotli.org) | [GraphEdge](Graph.md#edges) |
+| `privateConnectionPRIds` | 1.2 | NONE | NONE | [PRId](Graph.md#pseudonymous-relationship-identifiers) |
+
+Data for each data type is initially formatted as a stream of Avro objects that should conform to the schema specified.
+Avro file- and block-level information (including in-stream schema) is omitted.
+The Avro stream is then compressed (or not) and encrypted (or not) as specified.
+
+`curve25519xsalsa20poly1305` (that is, X25519 key exchange, XSalsa20 encryption, and Poly1305 message authentication) is the default authenticated encryption algorithm used in the [NaCl](https://nacl.cr.yp.to) ("Salt") library, and its successor [libsodium](https://libsodium.org). In the specification of cryptographic operations below, relevant methods from these libraries are noted. While these specific libraries are not required for DSNP compatibility, they are highly recommended.
+
+## Data Chunks
+
+Because blockchain systems often have specific limits to the amount of data that can be included in a given transaction, operations on user data deal with the data in discrete chunks.
+As implementation strategies may vary, implementations MUST define their own maximum chunk size in bytes to be used in the operations described below.
+
+## Entity Tags
+
+To allow for application-layer caching and prevent race conditions between different applications attempting to update the same data, an implementation MUST provide a non-empty `etag` (entity tag) value with each chunk fetched, and update this for each chunk replaced.
+
+An `etag` is any ASCII-encoded string with a minimum length of one byte and a maximum length of 255 bytes.
+
+An implementation MUST fail the Replace User Data Operation if the `etag` values supplied do not match the current `etag` values for all chunks of the specified data type.
+
+## Replace User Data Operation
+
+The Replace User Data Operation takes the following parameters:
```

Comment:
Need to add a key reference/index to help establish the current `keyAgreement` key used to aid in later decryption (here and in Get User Data).